File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/CASMobileAdsViewFactory.kt
package com.cleveradssolutions.plugin.flutter

import android.content.Context
import android.graphics.Color
import android.util.Log
import android.view.View
import android.widget.TextView
import io.flutter.BuildConfig
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

internal class CASMobileAdsViewFactory(
    private val manager: AdInstanceManager
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context?, viewId: Int, args: Any?): PlatformView {
        val argsMap = args as? Map<*, *>
        var adId = 0
        if (argsMap != null) {
            adId = argsMap["adId"] as Int
            val width = argsMap["width"] as Int
            val height = argsMap["height"] as Int

            val flutterAd = manager.findAd(adId)
            val adView = flutterAd?.getPlatformView(width, height)
            if (adView != null) {
                return adView
            }
        }

        val message = "Ad View with the following id could not be found: $adId."
        Log.e("CAS.AI.Flutter", message)
        if (BuildConfig.DEBUG) {
            val errorView = TextView(context)
            errorView.text = message
            errorView.setBackgroundColor(Color.RED)
            errorView.setTextColor(Color.YELLOW)
            return FlutterPlatformView(errorView)
        } else {
            return FlutterPlatformView(View(context))
        }
    }
}

File: ios/Classes/CASMobileAdsViewFactory.swift
import Flutter

class CASMobileAdsViewFactory: NSObject, FlutterPlatformViewFactory{
    
    private let manager: AdInstanceManager
    
    init(manager: AdInstanceManager) {
        self.manager = manager
        super.init()
    }
    
    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        var adId = 0
        
        if let argsMap = args as? [String: Any] {
            adId = argsMap["adId"] as! Int
            let width = argsMap["width"] as! Int
            let height = argsMap["height"] as! Int
            
            if let flutterAd = manager.findAd(adId) as? FlutterPlatformView {
                if let nativeAd = flutterAd as? FlutterNativeAd{
                    nativeAd.updatePlatformView(width: CGFloat(width),
                                                height: CGFloat(height))
                }
                
                    return flutterAd
            }
        }

        let reason = "Ad View with the following id could not be found: \(adId)."
        NSException(name: .invalidArgumentException, reason: reason, userInfo: nil).raise()
        fatalError(reason)
    }
    
    func createArgsCodec() -> any FlutterMessageCodec & NSObjectProtocol {
        FlutterStandardMessageCodec.sharedInstance()
    }
}
